other/autoregressive-process/model.jl

```diff
@@ -0,0 +1,91 @@
+    proclen::Int         = 750                 # Process length
+    # Recurrent net parameters
+    dev                  = cpu                 # Device: cpu or gpu
+    opt                  = ADAM                # Optimizer
+    η::Float64           = 2e-3                # Learning rate
+    hidden_nodes::Int    = 64                  # Number of hidden nodes
+    hidden_layers::Int   = 2                   # Number of hidden layers
+    layer                = LSTM                # Type of layer, should be one of LSTM, GRU, RNN
+    epochs::Int          = 100                 # Number of epochs
+    seqlen::Int          = 10                  # Sequence length to use as input
+    seqshift::Int        = 10                  # Shift between sequences (see utils.jl)
+    train_ratio::Float64 = .7                  # Percentage of data in the train set
+    verbose::Bool        = true                # Whether we log the results during training or not
+end
+
+# Creates a model according to the pre-defined hyperparameters `args`
+function build_model(args)
+    Chain(
+        args.layer(1, args.hidden_nodes),
+        [args.layer(args.hidden_nodes, args.hidden_nodes) for _ ∈ 1:args.hidden_layers-1]...,
+        Dense(args.hidden_nodes, 1, identity)
+    ) |> args.dev
+end
+
+# Creates training and testing samples according to hyperparameters `args`
+function generate_train_test_data(args)
+    # Generate full AR process
+    data = generate_process(args.ϕ, args.proclen)
+    # Create input X and output y (series shifted by 1)
+    X, y = data[1:end-1], data[2:end]
+    # Split data into training and testing sets
+    idx = round(Int, args.train_ratio * length(X))
+    Xtrain, Xtest = X[1:idx], X[idx+1:end]
+    ytrain, ytest = y[1:idx], y[idx+1:end]
+    # Transform data to time series batches and return
+    map(x -> batch_timeseries(x, args.seqlen, args.seqshift) |> args.dev, 
+        (Xtrain, Xtest, ytrain, ytest))
+end
+
+# Trains and outputs the model according to the chosen hyperparameters `args`
+function train_model(args)
+    Random.seed!(args.seed)
+    # Create recurrent model
+    m = build_model(args)
+    # Get data
+    Xtrain, Xtest, ytrain, ytest = generate_train_test_data(args)
+
+    opt = args.opt(args.η)
+    θ = Flux.params(m) # Keep track of model parameters
+    function loss(x, y)
+        # Warm up recurrent model on first observation
+        m(x[1])
+        # Compute mean squared error loss on the rest of the sequence
+        mean(Flux.Losses.mse.([m(xᵢ) for xᵢ ∈ x[2:end]], y[2:end]))
+    end
+    # Training loop
+    for i ∈ 1:args.epochs
+        Flux.reset!(m) # Reset hidden state of the recurrent model
+        # Compute the gradients of the loss function
+        ∇ = gradient(θ) do
+            loss(Xtrain, ytrain)
+        end
+        Flux.update!(opt, θ, ∇) # Update model parameters
+        if args.verbose && i % 10 == 0 # Log results every 10 epochs
+            Flux.testmode!(m)
```

Comment:
```suggestion
```
This is a no-op since the model doesn't have any normalization layers (specifically dropout, batchnorm or groupnorm).